gui/src/app/Scripting/webR/useWebR.ts

```diff
@@ -0,0 +1,189 @@
+import { RefObject, useCallback, useEffect, useState } from "react";
+import { WebR } from "webr";
+import { InterpreterStatus } from "@SpScripting/InterpreterTypes";
+import { writeConsoleOutToDiv } from "@SpScripting/OutputDivUtils";
+
+const captureOutputOptions = {
+  withAutoprint: true,
+  captureStreams: false,
+  captureConditions: false,
+  captureGraphics: false,
+} as const;
+
+type useWebRProps = {
+  consoleRef: RefObject<HTMLDivElement>;
+  imagesRef?: RefObject<HTMLDivElement>;
+  onStatus: (status: InterpreterStatus) => void;
+  onData?: (data: any) => void;
+};
+
+type RunRProps = {
+  code: string;
+  spData?: Record<string, any>;
+};
+
+const useWebR = ({ imagesRef, consoleRef, onStatus, onData }: useWebRProps) => {
+  const [webR, setWebR] = useState<WebR | null>(null);
+
+  const loadWebRInstance = useCallback(async () => {
+    if (webR) {
+      return webR;
+    }
+
+    onStatus("loading");
+    await sleep(100); // let the UI update
+
+    const w = new WebR();
+    await w.init();
+
+    onStatus("installing");
+    await sleep(100); // let the UI update
+    await w.installPackages(["jsonlite", "posterior"]);
+
+    setWebR(w);
+    return w;
+  }, [onStatus, webR]);
+
+  useEffect(() => {
+    const closedHandle = { closed: false };
+    if (webR) outputLoop(webR, consoleRef, imagesRef, closedHandle);
+
+    return () => {
+      if (webR) {
+        console.log("closing webR!");
+        webR.close();
+        setWebR(null);
+        onStatus("idle");
+      }
+      closedHandle.closed = true;
+    };
+  }, [consoleRef, imagesRef, onStatus, webR]);
+
+  const run = useCallback(
+    async ({ code, spData }: RunRProps) => {
+      try {
+        const webR = await loadWebRInstance();
+        const shelter = await new webR.Shelter();
+        onStatus("running");
+        await sleep(100); // let the UI update
+
+        try {
+          const globals: { [key: string]: any } = {
+            ".stan_playground": true,
+          };
+          if (spData) {
+            const spDataR = {
+              ...spData,
+              draws: await new shelter.RDouble(spData.draws.flat()),
+            };
+            globals[".SP_DATA_IN"] = await new shelter.RList(spDataR);
+          }
+
+          const env = await new shelter.REnvironment(globals);
+
+          const options = { ...captureOutputOptions, env };
+
+          // setup
+          await webR.evalRVoid(
+            `webr::shim_install()
+webr::canvas()`,
```

Comment:
These might be a hair more legible if you throw them into a const somewhere with less indentation & give them a name.